File: app/src/main/java/com/ywwynm/everythingdone/fragments/AlertDialogFragment.java
package com.ywwynm.everythingdone.fragments;

import android.app.Activity;
import android.app.DialogFragment;
import android.content.DialogInterface;
import android.graphics.Color;
import android.os.Bundle;
import androidx.core.content.ContextCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.TextView;

import com.ywwynm.everythingdone.R;
import com.ywwynm.everythingdone.utils.DisplayUtil;

/**
 * Created by ywwynm on 2015/10/9.
 * A subclass of {@link DialogFragment} to show alert information.
 */
public class AlertDialogFragment extends BaseDialogFragment {

    public static final String TAG = "AlertDialogFragment";

    private int[] mColors = new int[] { Color.BLACK, 0, Color.BLACK };

    private String mTitle;
    private String mContent;

    private String mConfirmText;
    private String mCancelText;

    private boolean mShowCancel = true;
    private ConfirmListener mConfirmListener;
    private CancelListener mCancelListener;
    private boolean mConfirmed = false;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);

        Activity activity = getActivity();
        if (mColors[1] == 0) {
            int contentColor = ContextCompat.getColor(activity, R.color.black_54p);
            mColors[1] = contentColor;
        }

        TextView tvTitle       = f(R.id.tv_title_alert);
        TextView tvContent     = f(R.id.tv_content_alert);
        TextView tvConfirmAsBt = f(R.id.tv_confirm_as_bt_alert);
        TextView tvCancelAsBt  = f(R.id.tv_cancel_as_bt_alert);

        if (mTitle != null) {
            tvTitle.setText(mTitle);
            applyAccent(tvTitle, mTitleBg, mColors[0]);
        } else {
            tvTitle.setVisibility(View.GONE);
        }

        if (mContent != null) {
            tvContent.setTextColor(mColors[1]);
            tvContent.setText(mContent);
        } else {
            tvContent.setVisibility(View.GONE);
        }

        if (mConfirmText != null) {
            tvConfirmAsBt.setText(mConfirmText);
        }
        applyAccent(tvConfirmAsBt, mConfirmBg, mColors[2]);
        tvConfirmAsBt.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mConfirmListener != null) {
                    mConfirmListener.onConfirm();
                }
                mConfirmed = true;
                dismiss();
            }
        });

        if (mShowCancel) {
            if (mCancelText != null) {
                tvCancelAsBt.setText(mCancelText);
            }
            tvCancelAsBt.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    dismiss();
                }
            });
        } else {
            tvCancelAsBt.setVisibility(View.GONE);
        }

        return mContentView;
    }

    @Override
    protected int getLayoutResource() {
        return R.layout.fragment_alert;
    }

    @Override
    public void onResume() {
        super.onResume();
        float screenDensity = DisplayUtil.getScreenDensity(getActivity());
        Window window = getDialog().getWindow();
        if (window != null) {
            window.setLayout((int) (screenDensity * 320), WindowManager.LayoutParams.WRAP_CONTENT);
        }
    }

    @Override
    public void onDismiss(DialogInterface dialog) {
        if (!mConfirmed && mCancelListener != null) {
            mCancelListener.onCancel();
        }
        mConfirmListener = null;
        mCancelListener  = null;
        super.onDismiss(dialog);
    }

    /**
     * Phase 8: paint {@code tv} with {@code bg} if non-null (PURE → single int,
     * GRADIENT → glyph-fill shader via {@link BackgroundUtil#applyTextBackground}).
     * Falls back to {@code fallbackColor} (the legacy int setter result) when
     * {@code bg} is null.
     */
    private void applyAccent(TextView tv, com.ywwynm.everythingdone.model.ThingBackground bg,
                             int fallbackColor) {
        if (bg != null) {
            com.ywwynm.everythingdone.utils.BackgroundUtil.applyTextBackground(tv, bg);
        } else {
            tv.setTextColor(fallbackColor);
        }
    }

    /** Phase 8: backing fields for the ThingBackground-typed setters below. */
    private com.ywwynm.everythingdone.model.ThingBackground mTitleBg;
    private com.ywwynm.everythingdone.model.ThingBackground mConfirmBg;

    public void setTitleBackground(com.ywwynm.everythingdone.model.ThingBackground bg) {
        mTitleBg = bg;
        if (bg != null) mColors[0] = bg.representativeColor();
    }

    public void setConfirmBackground(com.ywwynm.everythingdone.model.ThingBackground bg) {
        mConfirmBg = bg;
        if (bg != null) mColors[2] = bg.representativeColor();
    }

    public void setTitleColor(int color) {
        mTitleBg = null;
        mColors[0] = color;
    }

    public void setContentColor(int color) {
        mColors[1] = color;
    }

    public void setConfirmColor(int color) {
        mConfirmBg = null;
        mColors[2] = color;
    }

    public void setTitle(String title) {
        mTitle = title;
    }

    public void setContent(String content) {
        mContent = content;
    }

    public void setConfirmText(String confirmText) {
        mConfirmText = confirmText;
    }

    public void setConfirmListener(ConfirmListener listener) {
        mConfirmListener = listener;
    }

    public void setCancelText(String cancelText) {
        mCancelText = cancelText;
    }

    public void setCancelListener(CancelListener listener) {
        mCancelListener = listener;
    }

    public void setShowCancel(boolean showCancel) {
        mShowCancel = showCancel;
    }

    public interface ConfirmListener {
        void onConfirm();
    }

    public interface CancelListener {
        void onCancel();
    }
}

File: app/src/main/java/com/ywwynm/everythingdone/fragments/AlertDialogFragment.kt
@file:Suppress("DEPRECATION", "OVERRIDE_DEPRECATION")

package com.ywwynm.everythingdone.fragments

import android.content.DialogInterface
import android.os.Bundle
import androidx.core.content.ContextCompat
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.view.WindowManager
import android.widget.TextView

import com.ywwynm.everythingdone.R
import com.ywwynm.everythingdone.model.ThingBackground
import com.ywwynm.everythingdone.utils.BackgroundUtil
import com.ywwynm.everythingdone.utils.DisplayUtil

/**
 * Created by ywwynm on 2015/10/9.
 * Translated to Kotlin by ywwynm and Claude Opus 4.7 on 2026/5/20.
 * A subclass of DialogFragment to show alert information.
 */
open class AlertDialogFragment : BaseDialogFragment() {

    private var mColors: IntArray = intArrayOf(0, 0, 0)

    private var mTitle: String? = null
    private var mContent: String? = null

    private var mConfirmText: String? = null
    private var mCancelText: String? = null

    private var mShowCancel: Boolean = true
    private var mConfirmListener: ConfirmListener? = null
    private var mCancelListener: CancelListener? = null
    private var mConfirmed: Boolean = false

    /** Phase 8: backing fields for the ThingBackground-typed setters. */
    private var mTitleBg: ThingBackground? = null
    private var mConfirmBg: ThingBackground? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View? {
        super.onCreateView(inflater, container, savedInstanceState)

        val activity = activity!!
        if (mColors[0] == 0) {
            mColors[0] = ContextCompat.getColor(activity, R.color.app_chrome_on_surface_strong)
        }
        if (mColors[1] == 0) {
            val contentColor = ContextCompat.getColor(activity, R.color.app_chrome_on_surface_secondary)
            mColors[1] = contentColor
        }
        if (mColors[2] == 0) {
            mColors[2] = ContextCompat.getColor(activity, R.color.app_chrome_on_surface_strong)
        }

        val tvTitle: TextView       = f(R.id.tv_title_alert)!!
        val tvContent: TextView     = f(R.id.tv_content_alert)!!
        val tvConfirmAsBt: TextView = f(R.id.tv_confirm_as_bt_alert)!!
        val tvCancelAsBt: TextView  = f(R.id.tv_cancel_as_bt_alert)!!

        if (mTitle != null) {
            tvTitle.text = mTitle
            applyAccent(tvTitle, mTitleBg, mColors[0])
        } else {
            tvTitle.visibility = View.GONE
        }

        if (mContent != null) {
            tvContent.setTextColor(mColors[1])
            tvContent.text = mContent
        } else {
            tvContent.visibility = View.GONE
        }

        if (mConfirmText != null) {
            tvConfirmAsBt.text = mConfirmText
        }
        applyAccent(tvConfirmAsBt, mConfirmBg, mColors[2])
        tvConfirmAsBt.setOnClickListener {
            if (mConfirmListener != null) {
                mConfirmListener!!.onConfirm()
            }
            mConfirmed = true
            dismiss()
        }

        if (mShowCancel) {
            if (mCancelText != null) {
                tvCancelAsBt.text = mCancelText
            }
            tvCancelAsBt.setOnClickListener { dismiss() }
        } else {
            tvCancelAsBt.visibility = View.GONE
        }

        return mContentView
    }

    override fun getLayoutResource(): Int = R.layout.fragment_alert

    override fun onResume() {
        super.onResume()
        val screenDensity: Float = DisplayUtil.getScreenDensity(activity)
        val window: Window? = dialog!!.window
        window?.setLayout((screenDensity * 320).toInt(), WindowManager.LayoutParams.WRAP_CONTENT)
    }

    override fun onDismiss(dialog: DialogInterface) {
        if (!mConfirmed && mCancelListener != null) {
            mCancelListener!!.onCancel()
        }
        mConfirmListener = null
        mCancelListener  = null
        super.onDismiss(dialog)
    }

    /** Phase 8: paint `tv` with `bg` if non-null; falls back to `fallbackColor` otherwise. */
    private fun applyAccent(tv: TextView, bg: ThingBackground?, fallbackColor: Int) {
        if (bg != null) {
            BackgroundUtil.applyTextBackground(tv, bg)
        } else {
            tv.setTextColor(fallbackColor)
        }
    }

    open fun setTitleBackground(bg: ThingBackground?) {
        mTitleBg = bg
        if (bg != null) mColors[0] = bg.representativeColor()
    }

    open fun setConfirmBackground(bg: ThingBackground?) {
        mConfirmBg = bg
        if (bg != null) mColors[2] = bg.representativeColor()
    }

    open fun setTitleColor(color: Int) {
        mTitleBg = null
        mColors[0] = color
    }

    open fun setContentColor(color: Int) {
        mColors[1] = color
    }

    open fun setConfirmColor(color: Int) {
        mConfirmBg = null
        mColors[2] = color
    }

    open fun setTitle(title: String?) {
        mTitle = title
    }

    open fun setContent(content: String?) {
        mContent = content
    }

    open fun setConfirmText(confirmText: String?) {
        mConfirmText = confirmText
    }

    open fun setConfirmListener(listener: ConfirmListener?) {
        mConfirmListener = listener
    }

    open fun setCancelText(cancelText: String?) {
        mCancelText = cancelText
    }

    open fun setCancelListener(listener: CancelListener?) {
        mCancelListener = listener
    }

    open fun setShowCancel(showCancel: Boolean) {
        mShowCancel = showCancel
    }

    interface ConfirmListener {
        fun onConfirm()
    }

    interface CancelListener {
        fun onCancel()
    }

    companion object {
        const val TAG: String = "AlertDialogFragment"
    }
}
